### PR TITLE
Implement cursors for `BTreeSet`

### PR DESCRIPTION
**Pull request by kmicklas**
Tracking issue: https://github.com/rust-lang/rust/issues/107540

This is a straightforward wrapping of the map API, except that map's `CursorMut` does not make sense, because there is no value to mutate. Hence, map's `CursorMutKey` is wrapped here as just `CursorMut`, since it's unambiguous for sets and we don't normally speak of "keys". On the other hand, I can see some potential for confusion with `CursorMut` meaning different things in each module. I'm happy to take suggestions to improve that.

r? @Amanieu

### CODE REVIEW

**Comment by GrigorenkoPV** [comment]
Why is it unsafe to get a `CursorMut`? Shouldn't it be completely safe to get a `CursorMut` and use it to remove entries? Maybe the `CursorMut -> &mut T` methods should be marked `unsafe` instead?

**Comment by kmicklas** [comment]
> Why is it unsafe to get a `CursorMut`? Shouldn't it be completely safe to get a `CursorMut` and use it to remove entries? Maybe the `CursorMut -> &mut T` methods should be marked `unsafe` instead?

That's a good question. I was just propagating the `unsafe`ness from `with_mutable_key`, but I think the same argument applies to `BtreeMap`'s `CursorMutKey`.

**Comment by GrigorenkoPV** [comment]
> but I think the same argument applies to `BtreeMap`'s `CursorMutKey`.

Not really. From what I can tell, the Map has 3 kinds of cursors:
1. `Cursor`: view things
2. `CursorMut`: same as above + modify values, add/remove entries
3. `CursorMutKey`: same as above + modify keys

This PR seems to implement 1 & 3 for the Set, but not 2.

I am not aware why the decision was made to allow mutating a map's keys via an `unsafe` conversion to a separate cursor and not via `unsafe` methods on `CursorMut` itself, but I suppose we should preserve the same trichotomy for the Set as well (with some minor alterations, of course):
1. `Cursor`: view things
2. `CursorMut`: same as above + ~modify values,~ add/remove entries
3. ~`CursorMutKey`~ `CursorMutElem`: same as above + modify ~keys~ elements

OR (if we want only 2 cursors here) we should make the conversion to `CursorMut` safe and provide some `unsafe` element-mutating methods to it.

Disclaimer: I am not a t-libs (or t-libs-api) member.

**Comment by kmicklas** [comment]
> I am not aware why the decision was made to allow mutating a map's keys via an `unsafe` conversion to a separate cursor and not via `unsafe` methods on `CursorMut` itself

Personally that seems like the nicest API to me. (And is was what I was going for ignoring the accidental `unsafe` on the `CursorMut` construction methods.)

> but I suppose we should preserve the same trichotomy for the Set as well (with some minor alterations, of course):
> 
>     1. `Cursor`: view things
> 
>     2. `CursorMut`: same as above + ~modify values,~ add/remove entries
> 
>     3. ~`CursorMutKey`~ `CursorMutElem`: same as above + modify ~keys~ elements

Yeah this also makes sense and is most consistent with the map versions.

I'll wait for @Amanieu to give some input.

File: library/alloc/src/collections/btree/set.rs [hunk 1]
```diff
@@ -1817,5 +1990,257 @@ impl<'a, T: Ord> Iterator for Union<'a, T> {
...
+        self.inner.remove_prev().map(|(k, _)| k)
+    }
+}
+
+/// Error type returned by [`CursorMut::insert_before`] and
+/// [`CursorMut::insert_after`] if the element being inserted is not properly
+/// ordered with regards to adjacent elements.
+#[derive(Clone, PartialEq, Eq, Debug)]
+#[unstable(feature = "btree_cursors", issue = "107540")]
+pub struct UnorderedError {}
```
Comment (Amanieu) [inline library/alloc/src/collections/btree/set.rs:2226]: Can you reuse the one from `BTreeMap` here? It's fine if it mentions `Key` in the name, the meaning will be clear to anyone.

**Review by Amanieu — COMMENTED** [review]
There's 2 possible designs for exposing mutable keys to users. One through `CursorMutKey` and one through additional unsafe methods on `CusorMut`. The reason for choosing `CursorMutKey` was that it makes the API easier for users and avoids duplicating all next/prev methods on `CursorMut` with an unsafe version.

We may want to revisit this before stabilization, but for now the API in this PR looks fine and consistent with the one in `BTreeMap`.

**Comment by kmicklas** [comment]
> There's 2 possible designs for exposing mutable keys to users. One through `CursorMutKey` and one through additional unsafe methods on `CusorMut`. The reason for choosing `CursorMutKey` was that it makes the API easier for users and avoids duplicating all next/prev methods on `CursorMut` with an unsafe version.
> 
> We may want to revisit this before stabilization, but for now the API in this PR looks fine and consistent with the one in `BTreeMap`.

If I'm understanding this correctly, you'd prefer that we mirror the trichotomy with `CursorMutKey` then? I've updated the PR to exactly mirror the map API, which resolves the safety concerns too.

Since we're reusing `UnorderedKeyError` I left the name `CursorMutKey` as well, though I do like @GrigorenkoPV's suggestion of `CursorMutElem`. Happy to rename things.

**Comment by Amanieu** [comment]
No sorry, I meant to suggest keeping the dichotomy for `BTreeSet` as it is right now with just `Cursor` and `CursorMut`.

**Comment by kmicklas** [comment]
> No sorry, I meant to suggest keeping the dichotomy for `BTreeSet` as it is right now with just `Cursor` and `CursorMut`.

Ah okay. The original commit had no safe way to insert/remove elements with ordering checks. The map API has that and it seems important to retain. But if I make `lower_bound_mut`/`upper_bound_mut` safe, then I can't have the navigation methods (`next` etc.) return mutable references. So we'd need a suit of unsafe `next_mut`, `prev_mut`, etc. methods, and you said before this is what `CursorMutKey` was meant to avoid. What do you think?

**Comment by dongwq** [comment]
what a miss!! with this stable ASAP!!

**Comment by Amanieu** [comment]
Actually I like this new design even better than my original idea!

@bors r+

File: library/alloc/src/collections/btree/set.rs [hunk 2]
```diff
@@ -1183,6 +1183,178 @@ impl<T, A: Allocator + Clone> BTreeSet<T, A> {
...
+    /// let mut cursor = unsafe { set.upper_bound_mut(Bound::Excluded(&3)) };
+    /// assert_eq!(cursor.peek_prev(), Some(&2));
+    /// assert_eq!(cursor.peek_next(), Some(&3));
+    ///
+    /// let mut cursor = unsafe { set.upper_bound_mut(Bound::Unbounded) };
+    /// assert_eq!(cursor.peek_prev(), Some(&4));
+    /// assert_eq!(cursor.peek_next(), None);
+    /// ```
+    #[unstable(feature = "btree_cursors", issue = "107540")]
+    pub unsafe fn upper_bound_mut<Q: ?Sized>(&mut self, bound: Bound<&Q>) -> CursorMut<'_, T, A>
```
Comment (Mark-Simulacrum) [inline library/alloc/src/collections/btree/set.rs:1351]: Was it intended for this to be an unsafe fn? There's no Safety docs and lower_bound_mut is safe, so maybe just an oversight?

Comment (kmicklas) [reply library/alloc/src/collections/btree/set.rs:1351]: Yes this looks like an oversight!

**Pull request merged**